File: python/src/roax_canon/numbers.py
"""Canonical INTEGER and DECIMAL, over strings, never through a float.

This module is the reason the project exists (`docs/decisions.md` part 0, specification
section 6.4).
No value here is ever passed to :class:`float`, :class:`int` or :class:`decimal.Decimal`.
Digits are manipulated as text and the decimal point is moved by slicing, so a
forty-digit fraction and a nineteen-digit integer survive unchanged.

Three Python-specific traps are closed here rather than discovered later.

`$` in a Python regular expression also matches immediately before a trailing newline,
where JavaScript's matches only at end of input.
That divergence is already recorded in this repository: it made one reference
implementation accept ``"1.0\\n"`` and canonicalize it while the other rejected it
(`corpus/README.md`, "The divergence the two implementations caught").
Both grammars below are anchored ``\\A`` and ``\\Z``, which have no such behaviour.

`\\d` in a Python regular expression matches non-ASCII decimal digits by default, so
``١٢٣`` would satisfy a ``\\d+`` grammar and then fail or, worse, succeed at
``int()``, which also accepts fullwidth digits.
The grammars spell ``[0-9]`` out, exactly as specification section 6.2 writes them.

CPython 3.11 and later cap :func:`int` to :func:`str` conversion at 4300 digits by
default, so an exponent alone can raise :class:`ValueError` before any of this
specification's own bounds apply.
:func:`_exponent_value` refuses an absurd exponent by digit count before parsing it, so
the specification's 1024-digit bound is what rejects the input rather than an
interpreter setting.
"""

from __future__ import annotations

import re

from .errors import ErrorCode, GrammarError

__all__ = [
    "DIGIT_BOUND",
    "INTEGER_GRAMMAR",
    "DECIMAL_INPUT_GRAMMAR",
    "DECIMAL_OUTPUT_GRAMMAR",
    "canonical_integer",
    "canonical_decimal",
    "expanded_digit_count",
]

#: Specification section 6.2: a fixed constant of `ROAX-CANON/1`, not implementation
#: chosen.
#: Changing it changes which records canonicalize and is therefore a `canon` bump.
DIGIT_BOUND = 1024

#: Specification section 6.2, "Canonical integer".
INTEGER_GRAMMAR = re.compile(r"\A-?(0|[1-9][0-9]*)\Z")

#: Specification section 6.2, "Canonical decimal", input form.
#: This is exactly the FHIR R4 `decimal` pattern as shipped in the reference
#: `fhir/4.0.1/schema.json`, cited by path only.
DECIMAL_INPUT_GRAMMAR = re.compile(
    r"\A(?P<sign>-?)(?P<int>0|[1-9][0-9]*)(?:\.(?P<frac>[0-9]+))?"
    r"(?:[eE](?P<exp>[+-]?[0-9]+))?\Z"
)

#: Specification section 6.2, output form.
#: The same grammar `schemas/envelope-2.0.json` pins for a DECIMAL value.
DECIMAL_OUTPUT_GRAMMAR = re.compile(r"\A-?(0|[1-9][0-9]*)(\.[0-9]+)?\Z")

# An exponent with more decimal digits than this cannot bring the expanded form inside
# DIGIT_BOUND under any mantissa this grammar admits, so it is rejected without being
# parsed. 10 ** 7 already exceeds the bound by four orders of magnitude, in both
# directions: a positive exponent pads the integer part and a negative one pads the
# fraction.
_MAX_EXPONENT_DIGITS = 7


def _exponent_value(exp: str | None) -> int:
    """Parse the exponent, refusing one large enough to be a resource hazard.

    Returns 0 for an absent exponent, which is what specification section 6.2 means by
    "an absent exponent means ``e = 0``".
    """
    if exp is None:
        return 0
    sign = -1 if exp.startswith("-") else 1
    digits = exp.lstrip("+-").lstrip("0")
    if len(digits) > _MAX_EXPONENT_DIGITS:
        raise GrammarError(
            ErrorCode.DIGIT_BOUND_EXCEEDED,
            f"exponent of {len(digits)} digits expands past the {DIGIT_BOUND}-digit bound",
        )
    if not digits:
        return 0
    return sign * int(digits)


def expanded_digit_count(int_digits: str, frac_digits: str, exponent: int) -> int:
    """Digits the expanded positional form carries, computed without materializing it.

    Counts integer digits and fraction digits together, per specification section 6.2.

    **Two readings exist and this takes the literal one**, which is also the memory-safe
    one and the one both reference implementations took (`corpus/README.md`, ambiguity 2):
    the count is of the *padded* form, before the output-grammar normalization strips
    leading zeros.
    Under the other reading ``0e99999`` canonicalizes to ``0``; under this one it is
    rejected.
    No committed vector discriminates, and ``0e99999`` is not in the corpus.
    """
    total = len(int_digits) + len(frac_digits)
    point = len(int_digits) + exponent
    return max(point, 0) + max(total - point, 0)


def canonical_integer(text: str) -> str:
    """Canonical INTEGER form (tag 3), per specification section 6.2.

    Leading zeros are rejected, ``-0`` normalizes to ``0``, and anything else is an error.

    **The 1024-digit bound is applied here as well as to DECIMAL**, and that is a reading
    rather than a quotation.
    Specification section 6.2 states the bound under *Canonical decimal*, while its own
    justification paragraph counts "class 2's 40-digit integer" against it, which only
    makes sense if it governs INTEGER too (`corpus/README.md`, ambiguity 1).
    Both reference implementations read it the same way and no vector discriminates.
    """
    if not isinstance(text, str):
        raise GrammarError(
            ErrorCode.INTEGER_GRAMMAR,
            f"INTEGER value must be carried as a string, got {type(text).__name__}",
        )
    if INTEGER_GRAMMAR.match(text) is None:
        raise GrammarError(ErrorCode.INTEGER_GRAMMAR, f"not a canonical integer literal: {text!r}")
    digits = text[1:] if text.startswith("-") else text
    if len(digits) > DIGIT_BOUND:
        raise GrammarError(
            ErrorCode.DIGIT_BOUND_EXCEEDED,
            f"{len(digits)} digits exceeds the {DIGIT_BOUND}-digit bound",
        )
    if text == "-0":
        return "0"
    return text


def canonical_decimal(text: str) -> str:
    """Canonical DECIMAL form (tag 4), per specification section 6.2.

    Canonicalization does exactly two things: it expands exponent notation into
    positional notation, and it drops the sign of a zero-valued magnitude.
    The digit sequence is never rounded, extended or truncated.

    **Trailing zeros of the fraction are significant and survive.**
    ``0.010`` is not ``0.01``; FHIR R4 says SHALL.
    Trailing zeros of the integer part carry no precision in this grammar, so ``1e2``,
    ``1.0e2`` and ``100`` all encode to ``100`` and are the same leaf, while ``100.0`` is
    a different leaf.
    """
    if not isinstance(text, str):
        raise GrammarError(
            ErrorCode.DECIMAL_GRAMMAR,
            f"DECIMAL value must be carried as a string, got {type(text).__name__}",
        )
    m = DECIMAL_INPUT_GRAMMAR.match(text)
    if m is None:
        raise GrammarError(ErrorCode.DECIMAL_GRAMMAR, f"not a canonical decimal literal: {text!r}")

    sign = m.group("sign")
    int_digits = m.group("int")
    frac_digits = m.group("frac") or ""
    exponent = _exponent_value(m.group("exp"))

    if expanded_digit_count(int_digits, frac_digits, exponent) > DIGIT_BOUND:
        raise GrammarError(
            ErrorCode.DIGIT_BOUND_EXCEEDED,
            f"expanded positional form exceeds the {DIGIT_BOUND}-digit bound: {text!r}",
        )

    # Shift the decimal point right by `exponent` places over the concatenated digits,
    # padding with zeros wherever the point runs past the digits that are present.
    all_digits = int_digits + frac_digits
    point = len(int_digits) + exponent
    if point <= 0:
        int_part = ""
        frac_part = "0" * (-point) + all_digits
    elif point >= len(all_digits):
        int_part = all_digits + "0" * (point - len(all_digits))
        frac_part = ""
    else:
        int_part = all_digits[:point]
        frac_part = all_digits[point:]

    magnitude_is_zero = set(int_part + frac_part) <= {"0"}

    normalized_int = int_part.lstrip("0") or "0"
    out = normalized_int if frac_part == "" else f"{normalized_int}.{frac_part}"
    if sign and not magnitude_is_zero:
        out = "-" + out

    # A defect here would corrupt a root silently, so the output grammar is asserted
    # rather than assumed. This is the same grammar the envelope schema pins.
    assert DECIMAL_OUTPUT_GRAMMAR.match(out) is not None, out
    return out

File: src/numbers.ts
/**
 * Canonical numbers (specification section 6.2).
 *
 * Arbitrary precision, held as strings, never parsed into a machine integer and never through a
 * float. No digit of a record value is ever converted to a machine number here: every operation is
 * a string or `BigInt` operation over digit sequences, and `BigInt` appears only for exponent
 * bookkeeping, where it is exact.
 *
 * **There is exactly one `Number()` in this module and it converts no literal.** It narrows the
 * already-computed `BigInt` exponent AFTER the digit bound has proved `|e| <= 1024`, which makes
 * the narrowing exact; the comment at that line states the proof. Converting the exponent literal
 * directly would be the unsafe thing, for the reason given where it is parsed - a large exponent
 * saturates to `Infinity` and turns a rejection into an unbounded allocation.
 */

import { fail } from './errors.js';

/**
 * The total-digit bound of `ROAX-CANON/1`, counting integer and fraction digits together.
 *
 * A fixed constant of the canonicalization version and NOT implementation-chosen
 * (specification section 6.2). Section 13.3 rejects RDFC-1.0 partly because an
 * implementation-chosen iteration limit makes two conformant implementations disagree about which
 * documents they will canonicalize at all, and an implementation-chosen digit limit here would be
 * that same defect.
 */
export const MAX_TOTAL_DIGITS = 1024n;

/**
 * Grammars are anchored `^...$` with `\n` excluded explicitly rather than relying on the anchor.
 *
 * JavaScript's `$` matches only at end of input, so `"1.0\n"` is rejected here without further
 * work - but the same expression in a Python-family dialect ALSO matches before a trailing
 * newline and accepts it. `corpus/README.md` records that divergence as the one the two reference
 * implementations caught, and `reject-decimal-trailing-newline` pins it. The anchors are written
 * out so a reader can see the property is intended rather than inherited from the dialect.
 */
const INTEGER_GRAMMAR = /^-?(?:0|[1-9][0-9]*)$/;
const DECIMAL_INPUT_GRAMMAR = /^(-?)(0|[1-9][0-9]*)(?:\.([0-9]+))?(?:[eE]([+-]?[0-9]+))?$/;

function assertNoLineTerminator(input: string, code: 'integer-grammar' | 'decimal-grammar'): void {
  // Defence in depth against a future edit switching to a multiline-flagged expression, and
  // against a dialect whose `$` is lenient. A newline anywhere in a numeric literal is an error.
  if (/[\n\r\u2028\u2029]/.test(input)) {
    fail(code, 'a numeric literal may not contain a line terminator');
  }
}

/**
 * Canonicalizes an INTEGER (tag 3).
 *
 * - Grammar `^-?(0|[1-9][0-9]*)$`; leading zeros are rejected.
 * - `-0` normalizes to `0`.
 *
 * The 1024-digit bound is applied here as well as to DECIMAL. **This is a stated ambiguity, not a
 * settled reading.** Specification section 6.2 places the bound under *Canonical decimal*, while
 * the paragraph justifying it counts "class 2's 40-digit integer" against it, which only makes
 * sense if it governs INTEGER too. `corpus/README.md` records the same ambiguity as its first
 * finding and notes that both reference implementations apply it to INTEGER and that no vector
 * discriminates. This implementation takes the same reading, independently, for the reason the
 * corpus gives: an unbounded INTEGER is an unbounded allocation on hostile input.
 */
export function canonicalizeInteger(input: string): string {
  assertNoLineTerminator(input, 'integer-grammar');
  if (!INTEGER_GRAMMAR.test(input)) {
    fail('integer-grammar', `not a canonical integer: ${JSON.stringify(input)}`, { input });
  }
  const negative = input.startsWith('-');
  const digits = negative ? input.slice(1) : input;
  if (BigInt(digits.length) > MAX_TOTAL_DIGITS) {
    fail(
      'digit-bound-exceeded',
      `an integer of ${digits.length} digits exceeds the ${MAX_TOTAL_DIGITS} digit bound`,
      { digits: digits.length },
    );
  }
  // A zero magnitude loses its sign. `-0` and `0` are the same value and must be the same leaf.
  if (digits === '0') {
    return '0';
  }
  return negative ? `-${digits}` : digits;
}

/**
 * Canonicalizes a DECIMAL (tag 4).
 *
 * Canonicalization does exactly two things (specification section 6.2): it expands exponent
 * notation into positional notation, and it drops the sign of a zero-valued magnitude. The digit
 * sequence is never rounded, extended or truncated to a target precision, so a trailing zero of
 * the FRACTION survives and `0.010` is not `0.01` - which FHIR R4 states as a SHALL.
 *
 * This is the line dogtag takes the other way: `dogtag-mono-repo`,
 * `crates/dogtag-standard-rs/src/encode.rs:51-59` strips trailing zeros. Porting that function
 * without changing those lines would break FHIR conformance silently.
 */
export function canonicalizeDecimal(input: string): string {
  assertNoLineTerminator(input, 'decimal-grammar');
  const m = DECIMAL_INPUT_GRAMMAR.exec(input);
  if (m === null) {
    fail('decimal-grammar', `not a canonical decimal: ${JSON.stringify(input)}`, { input });
  }
  const negative = m[1] === '-';
  const intDigits = m[2] ?? '';
  const fracDigits = m[3] ?? '';
  // Exponent digits are read with BigInt, exactly. `Number()` would saturate a large exponent to
  // Infinity and turn a rejection into an unbounded allocation.
  const e = m[4] === undefined ? 0n : BigInt(m[4]);

  const i = BigInt(intDigits.length);
  const f = BigInt(fracDigits.length);

  // The digit count of the expanded positional form, computed BEFORE materializing it so a
  // hostile exponent cannot allocate. Derived from the shift rules below:
  //   e >= 0: intLen = i + e,          fracLen = max(0, f - e)   -> total = i + max(e, f)
  //   e <  0: k = -e; if k <= i  intLen = i - k, fracLen = f + k -> total = i + f
  //                   if k >  i  intLen = 0,     fracLen = k + f -> total = k + f
  //           which is total = max(i, k) + f
  const totalDigits = e >= 0n ? i + (e > f ? e : f) : (i > -e ? i : -e) + f;

  // **What the bound counts is a stated ambiguity, and this is the literal reading.**
  // Specification section 6.2 bounds "the expanded positional form", which is taken here as the
  // padded form BEFORE the output-grammar normalization below. `corpus/README.md` records the
  // same ambiguity as its second finding: under this reading `0e99999` is rejected, and under the
  // other it canonicalizes to `0`. No committed vector carries `0e99999`, so nothing
  // discriminates. This reading is chosen for the reason the corpus gives - it is the literal one
  // and the memory-safe one - and it is pinned at the edge by `decimal-at-digit-bound` (`1e1023`,
  // 1024 digits, accepted) and `reject-decimal-just-over-digit-bound` (`1e1024`, 1025, rejected).
  if (totalDigits > MAX_TOTAL_DIGITS) {
    fail(
      'digit-bound-exceeded',
      `the expanded positional form would carry ${totalDigits} digits, ` +
        `above the ${MAX_TOTAL_DIGITS} digit bound`,
      { digits: totalDigits.toString() },
    );
  }

  // Past the bound check the shift is safe to materialize. The bound also constrains the exponent
  // itself: `totalDigits >= e` when `e >= 0` and `totalDigits >= -e` when `e < 0`, so surviving
  // the check means `|e| <= 1024` and this narrowing is exact rather than a float conversion of a
  // literal.
  const shift = Number(e);
  let intPart: string;
  let fracPart: string;
  if (shift >= 0) {
    if (shift <= fracDigits.length) {
      intPart = intDigits + fracDigits.slice(0, shift);
      fracPart = fracDigits.slice(shift);
    } else {
      // The point runs past the digits that are present, so the gap pads with `0`.
      intPart = intDigits + fracDigits + '0'.repeat(shift - fracDigits.length);
      fracPart = '';
    }
  } else {
    const k = -shift;
    if (k <= intDigits.length) {
      intPart = intDigits.slice(0, intDigits.length - k);
      fracPart = intDigits.slice(intDigits.length - k) + fracDigits;
    } else {
      intPart = '';
      fracPart = '0'.repeat(k - intDigits.length) + intDigits + fracDigits;
    }
  }

  // Normalize to the output grammar `^-?(0|[1-9][0-9]*)(\.[0-9]+)?$`.
  //
  // Leading zeros of the integer part are removed and an emptied integer part becomes a single
  // `0`. A fraction of ZERO DIGITS is dropped along with its `.`; a fraction of zero-VALUED digits
  // is not, which is the whole point - `0.010` keeps `010` and `1.00e1` keeps the `0` of `10.0`.
  let normalizedInt = intPart.replace(/^0+/, '');
  if (normalizedInt === '') {
    normalizedInt = '0';
  }
  const isZeroMagnitude = /^0*$/.test(intPart) && /^0*$/.test(fracPart);
  // A zero-valued magnitude keeps its fraction digits and loses its sign, so `-0.00` is `0.00`.
  const sign = negative && !isZeroMagnitude ? '-' : '';
  return fracPart.length === 0 ? `${sign}${normalizedInt}` : `${sign}${normalizedInt}.${fracPart}`;
}
